packages/model-viewer/src/three-components/gltf-instance/VariantMaterialLoaderPlugin.ts

```diff
@@ -0,0 +1,275 @@
+/**
+ * Materials variants extension
+ *
+ * Specification:
+ * https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_variants
+ */
+
+import {Material, Mesh, Object3D} from 'three';
+import {GLTF, GLTFLoaderPlugin, GLTFParser} from 'three/examples/jsm/loaders/GLTFLoader.js';
+
+/**
+ * KHR_materials_variants specification allows duplicated variant names
+ * but it makes handling the extension complex.
+ * We ensure tha names and make it easier.
+ * If you want to export the extension with the original names
+ * you are recommended to write GLTFExporter plugin to restore the names.
+ *
+ * @param variantNames {Array<string>}
+ * @return {Array<string>}
+ */
+const ensureUniqueNames = (variantNames: string[]) => {
+  const uniqueNames = [];
+  const knownNames: {[key: string]: boolean} = {};
+
+  for (const name of variantNames) {
+    let uniqueName = name;
+    let suffix = 0;
+    // @TODO: An easy solution.
+    //        O(N^2) in the worst scenario where N is variantNames.length.
+    //        Fix me if needed.
+    while (knownNames[uniqueName] !== undefined) {
+      uniqueName = name + '.' + (++suffix);
+    }
+    knownNames[uniqueName] = true;
+    uniqueNames.push(uniqueName);
+  }
+
+  return uniqueNames;
+};
+
+/**
+ * Convert mappings array to table object to make handling the extension easier.
+ *
+ * @param
+ *     extensionDef {glTF.meshes[n].primitive.extensions.KHR_materials_variants}
+ * @param variantNames {Array<string>} Required to be unique names
+ * @return {Object}
+ */
+const mappingsArrayToTable = (extensionDef: any, variantNames: string[]) => {
+  const table:
+      {[key: string]:
+           {material: Material|null, gltfMaterialIndex: number}} = {};
+  for (const mapping of extensionDef.mappings) {
+    for (const variant of mapping.variants) {
+      table[variantNames[variant]] = {
+        material: null,
+        gltfMaterialIndex: mapping.material
+      };
+    }
+  }
+  return table;
+};
+
+/**
+ * @param object {THREE.Object3D}
+ * @return {boolean}
+ */
+const compatibleObject = (object: Object3D) => {
+  return 'material' in object &&  // easier than (!object.isMesh &&
+                                  // !object.isLine && !object.isPoints)
+      object!.userData &&         // just in case
+      object!.userData.variantMaterials;
+};
+
+export default class GLTFMaterialsVariantsExtension implements
+    GLTFLoaderPlugin {
+  parser: GLTFParser;
+  name: string;
+
+  constructor(parser: GLTFParser) {
+    this.parser = parser;
+    this.name = 'KHR_materials_variants';
+  }
+
+  // Note that the following properties will be overridden even if they are
+  // pre-defined
+  // - gltf.userData.variants
+  // - mesh.userData.variantMaterials
+  afterRoot(gltf: GLTF) {
+    const parser = this.parser;
+    const json = parser.json;
+
+    if (!json.extensions || !json.extensions[this.name]) {

```

Comment:
In this project's style we tend to explicitly put `json.extnsions == null` rather than checking truthiness.

Reply:
Done